Fix "Shipping methods not found" bug

foladipo opened the pull request:
Resolves #2265 .

There was a bug/misleading error that used to show up now and then whenever a user tried to checkout. This bug was due to a combination of:

- not showing a status message while shipping methods were being loaded.
- weak internet connections.
- a race-condition-like situation.

What this PR does is to show a loading icon + message while shipping methods are being retrieved from various APIs (e.g Shippo's). And if that retrieval fails, then this PR aims to show an appropriate error message too.

## Test instructions
1. Enable Shippo as your only shipping methods provider.
1. Open a new incognito browser window.
1. *As an anonymous/unregistered user*, add items to your cart.
1. Proceed to checkout.
1. Fill in your address and save it. You should now see a "Loading shipping methods... Please wait..." message in the shipping section.
1. Wait for a few seconds, and you'll see that the loading message in the preceding step is replaced by a list of shipping methods.

aaronjudd commented on imports/plugins/core/shipping/client/templates/checkout/shipping.html:
```diff
@@ -16,21 +16,32 @@ <h3 class="panel-title">
+            {{else}}
+              <div>
+                <h4>
+                  Loading shipping methods... Please wait...
```
This needs an i18n entry. 

@rymorgan some design direction here would be good.  I think I'd prefer to only see a loader.

I cringe to see this much Blaze updated without a React conversion.

aaronjudd commented on imports/plugins/included/shippo/server/methods/shippo.js:
```diff
@@ -410,6 +410,11 @@ export const methods = {
         apiKey
       });
+
+      if (_.isEqual({}, shippoShipment)) {
```
Use native ES6 over lodash.. `Object.is` or `Object.keys`
```
Object.keys(shippoShipment).length === 0
```

aaronjudd commented:
@foladipo if you'd like to pair program a little on this, as you get into the approach that @zenweasel and you have discussed, (with which) I mostly agree with the approach you have all discussed, but think it might end up getting complicated so if you need further discussion please reach out.

foladipo commented:
Thanks @aaronjudd . I will reach if I get stuck.

zenweasel commented:
I get this error as soon as I enter the checkout screen:

```
Exception in callback of async function: TypeError: Can't call method on  undefined
    at TypeError (native)
    at module.exports (/Users/brenthoover/.meteor/packages/ecmascript-runtime-server/.0.4.1.qqrjf3++os+web.browser+web.cordova/npm/node_modules/core-js/modules/_defined.js:3:28)
    at module.exports (/Users/brenthoover/.meteor/packages/ecmascript-runtime-server/.0.4.1.qqrjf3++os+web.browser+web.cordova/npm/node_modules/core-js/modules/_to-object.js:4:17)
    at Function.from (/Users/brenthoover/.meteor/packages/ecmascript-runtime-server/.0.4.1.qqrjf3++os+web.browser+web.cordova/npm/node_modules/core-js/modules/es6.array.from.js:14:19)
    at _toConsumableArray (/Users/brenthoover/Projects/js/reaction/.meteor/local/build/programs/server/app/app.js:32292:182)
    at getShippingRates (imports/plugins/included/shippo/server/hooks/rates.js:61:7)
    at server/api/hooks.js:54:14
    at Array.reduce (native)
    at Object.Hooks.Events.run (server/api/hooks.js:53:22)
    at [object Object].shippingGetShippingRates (server/methods/core/shipping.js:144:18)
    at packages/check.js:129:16
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
    at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:128:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at packages/ddp-server/livedata_server.js:1686:15
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
```

zenweasel commented:
If I change my address after getting the initial quotes I don't get the waiting spinner. Not sure how big of a deal it is but it would be great because you can't tell it's doing something.

foladipo commented on imports/plugins/included/shippo/server/hooks/rates.js:
```diff
@@ -3,7 +3,9 @@ import { Shipping, Packages } from "/lib/collections";
 // callback ran on getShippingRates hook
-function getShippingRates(rates, cart) {
+function getShippingRates(previousQueryResults, cart) {
+  // For some strange reason, the third parameter is always undefined.
```
This function is called when `Hooks.Events.run("onGetShippingRates", [rates, retrialTargets], cart)` is called. But some strange reason, if I run the above hook with four arguments, by the time it gets to `getShippingRates`, which should three arguments, the third one will be undefined. That's why I added those comments.

foladipo replied:
But I'll take them out though.

foladipo commented on imports/plugins/included/shipping-rates/server/hooks/hooks.js:
```diff
@@ -83,7 +84,10 @@ function getShippingRates(rates, cart) {
+  // This is the format expected by clients of this hook. The second
+  // element is supposed to be an array with details of any errors
+  // that occurred in this method.
+  return [rates, []];
```
Yes, I do agree that that will be better.

foladipo commented on imports/plugins/core/shipping/client/templates/checkout/shipping.js:
```diff
@@ -21,13 +29,34 @@ function cartShippingQuotes(currentCart) {
+  let queryStatus;
+
+  if (cart) {
+    if (cart.shipping) {
```
Ok. I'll do that.

zenweasel replied:
Ok, but you didn't make this change

foladipo commented on imports/plugins/core/shipping/client/templates/checkout/shipping.html:
```diff
@@ -16,21 +16,32 @@ <h3 class="panel-title">
+            {{#if hasShippingMethods}}
+              <div>
+                <h4>
+                  Loading shipping methods... Please wait...
```
Noted. I'll add one.

foladipo commented:
@zenweasel I have fixed the issues you raised. Namely:

1. `Error on first entry`: To reproduce this error, one has to be first be a guest user who is visiting the site for the first time. As proof, when you get to the checkout page, you should see the image below:

Then, as this guest user, you need to add items to your cart. Immediately you do, the error you reported would show up in the terminal. But I have resolved that now.

1. Now, if you change your address, you should see a loader, although it might come and go in a flash.

1. I've combined the conditions we discussed in `imports/plugins/core/shipping/client/templates/checkout/shipping.js`.

1. I've added translation entries for the text we talked about in `imports/plugins/core/shipping/client/templates/checkout/shipping.html`.

1. I've added JSDocs comments to `imports/plugins/included/shipping-rates/server/hooks/hooks.js`

1. The not so valuable comments that you pointed out have been removed.

zenweasel commented:
@foladipo Thanks. I will take a look at it first thing in the morning

zenweasel commented:
If I configure my shop for Flat Rate shipping and try to check out I get an endless spinner and rates never populate.

![cart_checkout](https://user-images.githubusercontent.com/72819/30137776-8ad4929a-9397-11e7-88f7-327db9e00cc5.png)

zenweasel commented:
I'm still not seeing the spinner when I update my shipping method

zenweasel commented:
If I enable Shippo and turn off my network (so the request fails) I get an error message in the console but there does not appear to be a retry and I never get an error message on the client side which is basically the main thing we were trying to accomplish with this PR.

rymorgan commented on imports/plugins/core/shipping/client/templates/checkout/shipping.html:
```diff
@@ -16,21 +16,30 @@ <h3 class="panel-title">
-                  </a>
+            {{#if hasShippingMethods}}
+              <div>
+                <h4 data-i18n="checkoutShipping.loadingShippingMethods"></h4>
```
@zenweasel I think our standard spinner works.

foladipo replied:
Ok. I'll take that text out.

foladipo commented on server/methods/core/shipping.js:
```diff
@@ -25,42 +25,115 @@ export const methods = {
+    if (cart && cart.shipping) {
+      update = {
+        $set: {
+          "shipping.0.shipmentQuotesQueryStatus": {
```
Ok. I used it because I tried using `"shipping.$.shipmentQuotesQueryStatus"` but it didn't work. What's the alternative?

zenweasel commented:
This appears to be working correctly now.

zenweasel commented:
@foladipo You have conflicts that need to be resolved here.

aaronjudd commented:
![cart_checkout](https://user-images.githubusercontent.com/439959/30387954-04da5992-9863-11e7-9aa1-cbb8bffe3fc2.png)
Tested and failed.

Basic test:  anonymous user, icognito, place an order, first time shipping methods are there. second time they are not. (same user, same session)

foladipo commented:
@aaronjudd I have fixed the error you observed.

aaronjudd commented:
Nice work! I believe this finally has resolved the bug the ghost bug I had been seeing for so long. 

@rymorgan I created https://github.com/reactioncommerce/reaction/issues/2863 which might some direction from you.  While this does PR does take care of the critical perfomance/loading issues, it doesn't resolve the user experience when switching an existing method and recalculating rates.
